nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/PutAzureDataLakeStorage.java

```diff
@@ -0,0 +1,121 @@
+            final String accountKey = context.getProperty(ACCOUNT_KEY).evaluateAttributeExpressions(flowFile).getValue();
+            final String sasToken = context.getProperty(PROP_SAS_TOKEN).evaluateAttributeExpressions(flowFile).getValue();
+            final String endpoint = String.format(Locale.ROOT, "https://%s.dfs.core.windows.net", accountName);
+            DataLakeServiceClient storageClient;
+            if (StringUtils.isNotBlank(accountKey)) {
+                final StorageSharedKeyCredential credential = new StorageSharedKeyCredential(accountName,
+                        accountKey);
+                storageClient = new DataLakeServiceClientBuilder().endpoint(endpoint).credential(credential)
+                        .buildClient();
+            } else if (StringUtils.isNotBlank(sasToken)) {
+                storageClient = new DataLakeServiceClientBuilder().endpoint(endpoint).sasToken(sasToken)
+                        .buildClient();
+            } else {
+                throw new IllegalArgumentException(String.format("Either '%s' or '%s' must be defined.",
+                        ACCOUNT_KEY.getDisplayName(), PROP_SAS_TOKEN.getDisplayName()));
+            }
+            final DataLakeFileSystemClient dataLakeFileSystemClient = storageClient.getFileSystemClient(fileSystem);
+            final DataLakeDirectoryClient directoryClient = dataLakeFileSystemClient.getDirectoryClient(directory);
+
+            final long length = flowFile.getSize();
+            final Map<String, String> attributes = new HashMap<>();
+            final DataLakeFileClient fileClient = directoryClient.createFile(fileName);
+            if (length > 0) {
+                try (final InputStream rawIn = session.read(flowFile); final BufferedInputStream in = new BufferedInputStream(rawIn)) {
+                    fileClient.append(in, 0, length);
+
+                } catch (final IOException | StorageErrorException e) {
+                    getLogger().error("Failed to create file. Reasons: " + e.getMessage());
+                    session.transfer(flowFile, REL_FAILURE);
+                }
+            }
+            attributes.put("azure.filesystem", fileSystem);
+            attributes.put("azure.filename", fileName);
+            attributes.put("azure.primaryUri", fileClient.getFileUrl().toString());
+            attributes.put("azure.length", String.valueOf(length));
+            fileClient.flush(length);
+
+            if (!attributes.isEmpty()) {
+                flowFile = session.putAllAttributes(flowFile, attributes);
+            }
+            session.transfer(flowFile, REL_SUCCESS);
+            final long transferMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNanos);
+            session.getProvenanceReporter().send(flowFile, fileClient.getFileUrl().toString(), transferMillis);
+        } catch (IllegalArgumentException e) {
+            getLogger().error("Failed to create {}, due to {}", new Object[]{flowFile, e}, e);

```

Comment:
"Failed to upload {FlowFile}" message may be more adequate. The FF already exists, so "Failed to create {FlowFile}" sounds a bit misleading for me.